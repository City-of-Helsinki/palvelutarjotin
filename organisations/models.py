from dateutil.relativedelta import relativedelta
from django.conf import settings
from django.contrib.auth import get_user_model
from django.contrib.auth.models import UserManager
from django.contrib.postgres.fields import ArrayField
from django.db import models, transaction
from django.db.models import Max, Q
from django.utils import timezone
from django.utils.translation import gettext_lazy as _
from django.utils.translation import pgettext_lazy
from helusers.models import AbstractUser

from common.models import TimestampedModel, UUIDPrimaryKeyModel
from organisations.services import (
    send_myprofile_creation_notification_to_admins,
    send_myprofile_organisations_accepted_notification,
)


class PersonQuerySet(models.QuerySet):
    def user_can_view(self, user):
        # Only return profile of logged in user if he's not staff
        if user.is_staff:
            return self
        elif user.is_authenticated:
            return self.filter(user=user)
        else:
            return self.none()

    def retention_period_exceeded(self):
        earliest_valid_timestamp = timezone.now() - relativedelta(
            months=settings.PERSONAL_DATA_RETENTION_PERIOD_MONTHS
        )

        all_enrolments_too_old = Q(max_enrolment_end_time=None) | Q(
            max_enrolment_end_time__lt=earliest_valid_timestamp
        )
        all_study_groups_too_old = Q(max_studygroup_end_time=None) | Q(
            max_studygroup_end_time__lt=earliest_valid_timestamp
        )

        return (
            # At least for now returns only unauthenticated Person's
            self.filter(user=None)
            .annotate(
                max_enrolment_end_time=Max("enrolment__occurrence__end_time"),
                max_studygroup_end_time=Max(
                    "studygroup__enrolments__occurrence__end_time"
                ),
            )
            .filter(all_enrolments_too_old & all_study_groups_too_old)
        )

    @transaction.atomic
    def delete(self):
        for obj in self:
            obj.set_related_objects_person_deleted_at()
        return super().delete()


class UserQueryset(models.QuerySet):
    @transaction.atomic
    def delete(self):
        for obj in self:
            obj.delete_related_p_event_contact_info()
        return super().delete()


class CustomUserManager(UserManager.from_queryset(UserQueryset)):
    pass


class User(AbstractUser):
    is_admin = models.BooleanField(_("admin status"), default=False)

    objects = CustomUserManager()

    # When creating an user, the name and the email can be left to blank.
    # In those cases, return username.
    def __str__(self):
        display_name = super().__str__()
        if not display_name:
            return self.username
        return display_name

    class Meta:
        verbose_name = _("user")
        verbose_name_plural = _("users")
        permissions = [
            (
                "can_administrate_user_permissions",
                "Can administrate user permissions",
            ),
        ]

    @transaction.atomic
    def delete(self, *args, **kwargs):
        self.delete_related_p_event_contact_info()
        return super().delete(*args, **kwargs)

    def delete_related_p_event_contact_info(self):
        if hasattr(self, "person"):
            self.person.p_event.delete_contact_info()


class Organisation(models.Model):
    TYPE_USER = "user"
    TYPE_PROVIDER = "provider"
    ORGANISATION_TYPES = (
        (TYPE_USER, _("User")),
        (TYPE_PROVIDER, _("Provider")),
    )
    name = models.CharField(max_length=255, verbose_name=_("name"))
    phone_number = models.CharField(
        verbose_name=_("phone number"), max_length=64, blank=True
    )
    type = models.CharField(
        choices=ORGANISATION_TYPES, verbose_name=_("type"), max_length=64
    )
    persons = models.ManyToManyField(
        "Person", verbose_name=_("persons"), related_name="organisations", blank=True
    )
    publisher_id = models.CharField(
        max_length=255, verbose_name=_("publisher id"), blank=True
    )

    class Meta:
        verbose_name = _("organisation")
        verbose_name_plural = _("organisations")

    def __str__(self):
        return f"{self.name} ({self.id})"

    def is_editable_by_user(self, user):
        return user.person.organisations.filter(id=self.id).exists()


class OrganisationProposal(models.Model):
    """
    When a member of a 3rd party organisation registers
    to the API from the providers UI, he can make a proposal to add
    a new 3rd party organisation.
    NOTE: Since the process is still quite unclear, the proposals and
    and 3rd party organisation wishes can be collected with this model
    and stored in database as detached. The use case is that an admin can
    see which organisation the new user likes to represent and then
    the real organisation.Organisation instance can be created
    (if necessary and not yet done) and linked to the user.
    """

    name = models.CharField(max_length=255, verbose_name=_("name"))
    description = models.CharField(
        max_length=255, verbose_name=_("description"), blank=True
    )
    phone_number = models.CharField(
        verbose_name=_("phone number"), max_length=64, blank=True
    )
    applicant = models.ForeignKey(
        "Person", verbose_name=_("applicant"), on_delete=models.CASCADE
    )

    class Meta:
        verbose_name = _("organisation proposal")
        verbose_name_plural = _("organisation proposals")

    def __str__(self):
        return f"{self.id} {self.name}"


class Person(UUIDPrimaryKeyModel, TimestampedModel):
    user = models.OneToOneField(
        get_user_model(),
        verbose_name=_("user"),
        on_delete=models.CASCADE,
        blank=True,
        null=True,
    )
    name = models.CharField(max_length=255, verbose_name=_("name"))
    phone_number = models.CharField(
        verbose_name=_("phone number"), max_length=64, blank=True
    )
    email_address = models.EmailField(max_length=255, verbose_name=_("email"))
    language = models.CharField(
        verbose_name=_("language"), max_length=10, default=settings.LANGUAGES[0][0]
    )
    place_ids = ArrayField(
        models.CharField(max_length=250),
        verbose_name=_("own places"),
        blank=True,
        default=list,
    )

    objects = PersonQuerySet.as_manager()

    class Meta:
        verbose_name = _("person")
        verbose_name_plural = _("persons")

    def __str__(self):
        username = self.user.username if self.user else None
        if username:
            return f"{self.name} ({username})"
        return f"{self.name}"

    @transaction.atomic
    def delete(self, *args, **kwargs):
        self.set_related_objects_person_deleted_at()
        return super().delete(*args, **kwargs)

    def set_related_objects_person_deleted_at(self):
        from occurrences.models import Enrolment, StudyGroup

        now = timezone.now()
        Enrolment.objects.filter(person=self).update(person_deleted_at=now)
        StudyGroup.objects.filter(person=self).update(person_deleted_at=now)

    def is_editable_by_user(self, user):
        return (
            user.person.organisations.get_queryset()
            .intersection(self.organisations.get_queryset())
            .exists()
        )

    def notify_myprofile_creation(self, custom_message=None):
        send_myprofile_creation_notification_to_admins(
            self,
            custom_message=custom_message,
        )

    def notify_myprofile_accepted(self, custom_message=None):
        send_myprofile_organisations_accepted_notification(
            self, custom_message=custom_message
        )


class EnrolleePersonalDataManager(models.Manager):
    def get_queryset(self):
        return super().get_queryset().filter(user=None)


class EnrolleePersonalData(Person):
    objects = EnrolleePersonalDataManager.from_queryset(PersonQuerySet)()

    class Meta:
        proxy = True
        # Move these under "occurrences" group in the admin UI. That is not a perfect
        # place for these either, but "organisations" would have been misleading
        # because these have nothing to do with organisations.
        app_label = "occurrences"
        verbose_name = pgettext_lazy("singular", "Enrollee personal data")
        verbose_name_plural = pgettext_lazy("plural", "Enrollee personal data")
