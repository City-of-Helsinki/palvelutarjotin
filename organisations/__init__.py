default_app_config = "organisations.apps.OrganisationsConfig"
