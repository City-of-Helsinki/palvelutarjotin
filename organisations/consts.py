class NotificationTemplate:
    MYPROFILE_CREATION = "person_myprofile_creation"
    MYPROFILE_ACCEPTED = "person_myprofile_accepted"
