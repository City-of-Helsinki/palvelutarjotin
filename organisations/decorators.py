from graphql_jwt.decorators import user_passes_test

event_staff_member_required = user_passes_test(
    lambda u: getattr(u, "is_event_staff", False)
)
