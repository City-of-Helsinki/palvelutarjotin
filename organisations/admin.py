from typing import Any

from auditlog_extra.mixins import AuditlogAdminViewAccessLogMixin
from django import forms
from django.conf import settings
from django.contrib import admin, messages
from django.contrib.admin.widgets import FilteredSelectMultiple
from django.contrib.auth import get_user_model
from django.contrib.auth.admin import UserAdmin as DjangoUserAdmin
from django.contrib.auth.forms import UserChangeForm
from django.db.models.query import QuerySet
from django.http import HttpRequest, HttpResponseRedirect
from django.urls import reverse
from django.utils.translation import gettext as _
from django_admin_listfilter_dropdown.filters import RelatedDropdownFilter

from occurrences.models import StudyGroup
from organisations.models import Organisation, OrganisationProposal, Person, User


@admin.register(Organisation)
class OrganisationAdmin(AuditlogAdminViewAccessLogMixin, admin.ModelAdmin):
    filter_horizontal = ("persons",)
    list_display = ("name", "type", "publisher_id")
    list_filter = ["type"]
    search_fields = ["name", "publisher_id"]
    exclude = ("id",)
    actions = (
        "export_organisation_persons",
        "export_organisation_persons_csv",
    )

    def export_organisation_persons(self, request, queryset):
        selected = queryset.values_list("pk", flat=True)
        return HttpResponseRedirect(
            "%s?ids=%s"
            % (
                reverse("report_organisation_persons", current_app="reports"),
                ",".join(str(pk) for pk in selected),
            )
        )

    def export_organisation_persons_csv(self, request, queryset):
        selected = queryset.values_list("pk", flat=True)
        return HttpResponseRedirect(
            "%s?ids=%s"
            % (
                reverse("report_organisation_persons_csv", current_app="reports"),
                ",".join(str(pk) for pk in selected),
            )
        )

    export_organisation_persons_csv.short_description = _(
        "Export organisation persons (csv)"
    )


class OrganisationInline(admin.TabularInline):
    model = Organisation.persons.through


class OrganisationProposalAdminForm(forms.ModelForm):
    class Meta:
        model = OrganisationProposal
        fields = "__all__"

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.fields["applicant"].queryset = Person.objects.all().order_by(
            "name", "-created_at"
        )
        self.fields["applicant"].label_from_instance = (
            lambda instance: f"{instance.__str__()} (id: {instance.id})"
        )


@admin.register(OrganisationProposal)
class OrganisationProposalAdmin(AuditlogAdminViewAccessLogMixin, admin.ModelAdmin):
    list_display = ("name", "applicant")
    search_fields = (
        "name",
        "applicant__name",
        "applicant__user__username",
    )
    form = OrganisationProposalAdminForm


class UserExistenceListFilter(admin.SimpleListFilter):
    """
    Does the instance have a user profile or not?
    List filter options: All,True (for have user profile), False (for doesn't have)
    """

    title = _("has user profile")
    parameter_name = "person-has-user"

    def lookups(self, request, model_admin):
        return (
            (True, _("True")),
            (False, _("False")),
        )

    def queryset(self, request, queryset):
        if self.value() == "True":
            return queryset.filter(user__isnull=False)
        if self.value() == "False":
            return queryset.filter(user__isnull=True)


class RetentionPeriodExceededListFilter(admin.SimpleListFilter):
    title = _("retention period exceeded")
    parameter_name = "retention-period-exceeded"

    def lookups(self, request, model_admin):
        return ((True, _("True")),)

    def queryset(self, request, queryset):
        if self.value() == "True":
            return queryset.retention_period_exceeded()


class RelatedNameDropdownFilter(RelatedDropdownFilter):
    def field_admin_ordering(self, field, request, model_admin):
        return ["name"]


class StudyGroupInline(admin.TabularInline):
    model = StudyGroup
    extra = 0
    show_change_link = True

    def has_change_permission(self, request, obj=None):
        return False


class PersonAdminForm(forms.ModelForm):
    language = forms.ChoiceField(choices=settings.LANGUAGES)

    class Meta:
        model = Person
        fields = "__all__"

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        user = self.fields["user"]
        # Show only users without any links to a person
        user.queryset = User.objects.filter(person__isnull=True)
        if self.instance.user:
            # If an user instance is already set, add it to queryset with an union.
            user.queryset |= User.objects.filter(pk=self.instance.user.pk)
            user.initial = self.instance.user
        user.help_text = _(
            "Only the users that aren't linked to a person yet, can be selected."
        )


@admin.register(Person)
class PersonAdmin(AuditlogAdminViewAccessLogMixin, admin.ModelAdmin):
    enable_list_view_audit_logging = True
    list_display = (
        "id",
        "name",
        "user",
        "email_address",
        "created_at",
        "updated_at",
        "get_organisation_names",
        "get_study_groups",
    )
    fields = ("user", "name", "phone_number", "email_address", "language", "place_ids")
    ordering = ("-created_at",)
    inlines = (OrganisationInline, StudyGroupInline)
    list_filter = [
        "created_at",
        UserExistenceListFilter,
        ("organisations", RelatedNameDropdownFilter),
        RetentionPeriodExceededListFilter,
    ]
    search_fields = ["name", "email_address"]
    form = PersonAdminForm
    actions = (
        "export_persons",
        "export_persons_csv",
    )

    def get_queryset(self, request: HttpRequest) -> QuerySet[Any]:
        return (
            super()
            .get_queryset(request)
            .prefetch_related("organisations")
            .prefetch_related("studygroup_set")
        )

    def get_organisation_names(self, obj):
        return ", ".join((org.name for org in obj.organisations.all()))

    get_organisation_names.short_description = _("organisations")

    def get_study_groups(self, obj):
        return ", ".join((str(group) for group in obj.studygroup_set.all()))

    get_study_groups.short_description = _("study groups")

    def export_persons(self, request, queryset):
        selected = queryset.values_list("pk", flat=True)
        return HttpResponseRedirect(
            "%s?ids=%s"
            % (
                reverse("report_persons", current_app="reports"),
                ",".join(str(pk) for pk in selected),
            )
        )

    def export_persons_csv(self, request, queryset):
        selected = queryset.values_list("pk", flat=True)
        return HttpResponseRedirect(
            "%s?ids=%s"
            % (
                reverse("report_persons_csv", current_app="reports"),
                ",".join(str(pk) for pk in selected),
            )
        )


class UserPersonInLine(admin.StackedInline):
    """
    UserPersonInLine is created to make a support visible and clickable
    view and navigation between user and person from an admin user change form.

    NOTE: If any of the fields are writable, users that shouldn't have
    a person instance linked to them, will get one, because when any of the
    fields in the (person) inline form are included in save submit request,
    a person instance will be created - this should be avoided, since we
    would also like to support users without persons.
    """

    model = Person
    verbose_name = _("Person")
    verbose_name_plural = _(
        "Persons (The Kultus provider model extension for the User model)"
    )
    fields = ["name", "email_address", "language"]
    readonly_fields = fields  # All fields should be readonly fields
    show_change_link = True  # show a link to a person instance
    can_delete = False  # Prevent the deletion of a person instance
    classes = ["collapse"]  # Collapsible inline form


class UserAdminForm(UserChangeForm):
    organisations = forms.ModelMultipleChoiceField(
        queryset=Organisation.objects.all(),
        required=False,
        widget=FilteredSelectMultiple(verbose_name="Organisations", is_stacked=False),
        help_text=_(
            "Select organisations to link them with the user's person relation. "
            + "The organisations selection is disabled "
            + "when the user is not linked to any person instance."
        ),
    )

    class Meta:
        model = User
        fields = "__all__"

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if self.instance.pk:
            try:
                self.fields[
                    "organisations"
                ].initial = self.instance.person.organisations.all()
            except Person.DoesNotExist:
                self.fields["organisations"].disabled = True

    def save(self, commit=True):
        user = super().save(commit=False)
        if commit:
            user.save()

        if user.pk and hasattr(user, "person"):
            user.person.organisations.set(self.cleaned_data["organisations"])
            self.save_m2m()

        return user


@admin.register(get_user_model())
class UserAdmin(AuditlogAdminViewAccessLogMixin, DjangoUserAdmin):
    enable_list_view_audit_logging = True
    fieldsets = (
        (None, {"fields": ("username", "password")}),
        (_("Personal info"), {"fields": ("first_name", "last_name", "email")}),
        (
            _("Permissions"),
            {
                "fields": (
                    "is_active",
                    "is_event_staff",
                    "organisation_proposals",
                    "organisations",
                ),
            },
        ),
        (
            _("Advanced permission settings"),
            {
                "fields": (
                    "is_admin",
                    "is_staff",
                    "is_superuser",
                    "groups",
                    "user_permissions",
                ),
                "classes": ["collapse in"],
            },
        ),
        (
            _("Other information"),
            {
                "fields": ("last_login", "date_joined", "uuid", "ad_groups"),
                "classes": ["collapse in"],
            },
        ),
    )
    list_display = (
        "username",
        "email",
        "first_name",
        "last_name",
        "is_event_staff",
        "is_staff",
        "has_person",
        "uuid",
        "date_joined",
        "last_login",
    )
    list_filter = (
        "date_joined",
        "last_login",
        "is_event_staff",
        "is_staff",
        "is_admin",
        "is_superuser",
        "is_active",
    )
    readonly_fields = (
        "last_login",  # "last_login" is a nice to know, but shouldn't be editable
        "date_joined",  # "date_joined" is a nice to know, but shouldn't be editable
        "uuid",
        "ad_groups",
        "organisation_proposals",
    )
    ordering = ("-date_joined",)
    date_hierarchy = "date_joined"
    # The person link is in huge role all around the system,
    # so it would be good to show the related information InLine.
    # All the fields needs to be read_only,
    # or otherwise a missing person relation is created on (every) save and
    # sometimes it is wanted to have a user without person, e.g a pure admin user
    # or a superuser.
    inlines = (UserPersonInLine,)
    form = UserAdminForm

    def save_form(self, request, form, change):
        """
        Given a ModelForm return an unsaved instance. ``change`` is True if
        the object is being changed, and False if it's being added.
        Overridden so the request could be combined with the model instance save
        and we could write a message about a mail is being sent.
        """

        original_user = form.instance
        is_notifiable_changes_done = False

        # When user is linked to a person...
        if self.has_person(original_user):
            is_notifiable_changes_done = self._has_organisations_changed(
                original_user.person, form
            ) or self._has_is_event_staff_changed(form)

        user = super().save_form(request, form, change)

        # And the organisations or is_event_staff status are changed...
        if is_notifiable_changes_done:
            # And the edited user is accepted as a provider,
            # since he is an event staff member...
            if getattr(user, "is_event_staff", False):
                # And he is linked to some organisations...
                if user.person.organisations.count() > 0:
                    # And the user instance is different than the current user..
                    if user.id != request.user.id:
                        self._notify_user_of_account_activation(request, form)

        # Return the return value of the original save_form -method.
        return user

    def get_readonly_fields(self, request, obj=None):
        """
        Staff (not superusers) should not manage perms of Users.
        """
        readonly_fields = super(UserAdmin, self).get_readonly_fields(request, obj)
        if not request.user.has_perm("organisations.can_administrate_user_permissions"):
            readonly_fields += (
                "username",
                "is_superuser",
                "groups",
                "user_permissions",
            )
        return readonly_fields

    def has_person(self, obj):
        try:
            return bool(obj.person)
        except Person.DoesNotExist:
            return False

    has_person.boolean = True

    def organisation_proposals(self, obj):
        if obj.person:
            organisation_proposals = obj.person.organisationproposal_set.all().order_by(
                "name"
            )
            return ", ".join([org.name for org in organisation_proposals])
        return None

    def _has_organisations_changed(self, person, form):
        return (
            bool(
                sorted(list(person.organisations.all().values_list("id", flat=True)))
                != sorted([org.id for org in form.cleaned_data["organisations"]])
            )
            if person
            else False
        )

    def _has_is_event_staff_changed(self, form):
        return bool("is_event_staff" in form.changed_data)

    def _notify_user_of_account_activation(self, request, form):
        user = form.instance

        # Send a mail to the accepted user
        user.person.notify_myprofile_accepted()

        # Info about the sent mail
        messages.add_message(
            request,
            messages.INFO,
            "An email is sent to the user about the organisation changes "
            + "and the profile acceptance!",
        )
