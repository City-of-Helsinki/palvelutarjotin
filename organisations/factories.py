import factory
from django.contrib.auth import get_user_model
from organisations.models import Organisation, OrganisationProposal, Person


class UserFactory(factory.django.DjangoModelFactory):
    first_name = factory.Faker("first_name")
    last_name = factory.Faker("last_name")
    username = factory.Faker("user_name")
    email = factory.Faker("email")

    class Meta:
        model = get_user_model()


class PersonFactory(factory.django.DjangoModelFactory):
    user = factory.SubFactory(UserFactory)
    name = factory.Faker("name")
    phone_number = factory.Faker("phone_number")
    email_address = factory.Faker("email")

    class Meta:
        model = Person

    @factory.post_generation
    def organisations(self, create, extracted, **kwargs):
        if not create:
            # Simple build, do nothing.
            return

        if extracted:
            # A list of organisations were passed in, use them
            for organisation in extracted:
                self.organisations.add(organisation)


class OrganisationFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    phone_number = factory.Faker("phone_number")
    type = factory.Faker(
        "random_element", elements=[t[0] for t in Organisation.ORGANISATION_TYPES],
    )
    publisher_id = factory.Faker("pystr", max_chars=5)

    class Meta:
        model = Organisation


class OrganisationProposalFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    description = factory.Faker("text", max_nb_chars=255)
    phone_number = factory.Faker("phone_number")
    applicant = factory.SubFactory(PersonFactory)

    class Meta:
        model = OrganisationProposal
