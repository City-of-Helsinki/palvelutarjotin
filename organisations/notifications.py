from django.utils.translation import gettext_lazy as _
from django_ilmoitin.dummy_context import dummy_context
from django_ilmoitin.registry import notifications

from organisations.consts import NotificationTemplate
from organisations.factories import (
    OrganisationFactory,
    OrganisationProposalFactory,
    PersonFactory,
)
from organisations.services import get_user_change_form_url, get_user_list_url

TEMPLATES = [
    (NotificationTemplate.MYPROFILE_CREATION, _("My profile creation")),
    (NotificationTemplate.MYPROFILE_ACCEPTED, _("My profile accepted")),
]
for template in TEMPLATES:
    notifications.register(template[0], template[1])

organisation = OrganisationFactory.build()
person = PersonFactory.build(organisations=[organisation])
organisation_proposal = OrganisationProposalFactory.build(applicant=person)

DEFAULT_DUMMY_CONTEXT = {
    "preview_mode": False,
    "person": person,
    "custom_message": "Custom message.",
}

dummy_context.update(
    {
        NotificationTemplate.MYPROFILE_CREATION: {
            **DEFAULT_DUMMY_CONTEXT,
            "user_change_form_url": get_user_change_form_url(person),
            "user_list_url": get_user_list_url(),
        },
        NotificationTemplate.MYPROFILE_ACCEPTED: DEFAULT_DUMMY_CONTEXT,
    }
)
