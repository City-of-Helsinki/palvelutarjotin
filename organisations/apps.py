from django.apps import AppConfig


class OrganisationsConfig(AppConfig):
    name = "organisations"

    def ready(self):
        import organisations.notifications  # noqa
