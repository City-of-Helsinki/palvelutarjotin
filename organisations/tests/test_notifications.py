import pytest
from django.core import mail
from organisations.factories import (
    OrganisationProposalFactory,
    PersonFactory,
    UserFactory,
)

from common.tests.utils import assert_mails_match_snapshot


@pytest.mark.django_db
def test_myprofile_creation_email(
    snapshot,
    notification_template_myprofile_creation_fi,
    notification_template_myprofile_creation_en,
):
    person = PersonFactory(id=123)
    OrganisationProposalFactory.create_batch(2, applicant=person)
    UserFactory.create_batch(2, is_admin=True)
    person.notify_myprofile_creation(custom_message="custom message")
    assert len(mail.outbox) == 2
    assert_mails_match_snapshot(snapshot)
