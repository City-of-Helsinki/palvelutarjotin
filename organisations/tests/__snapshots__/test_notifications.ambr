# serializer version: 1
# name: test_myprofile_accepted_email[en]
  list([
    '''
      no-reply@hel.ninja|['sreyes@example.com']|My profile accepted EN|
      <p>Hi Kyle Walls!</p>
      <p>Your account is now ready to be used in Kultus
      with the following organisations linked to your account:</p>
      <ul>
          <li>Graves and Sons</li>
          <li>Valdez-Thompson</li>
      </ul>
      
          <p>Custom message: custom message</p>
          
    ''',
  ])
# ---
# name: test_myprofile_accepted_email[fi]
  list([
    '''
      no-reply@hel.ninja|['sreyes@example.com']|My profile accepted FI|
      <p>Hei Kyle Walls!</p>
      <p>Sinun käyttäjäsi on nyt valmis käytettäväksi Kultuksessa
      seuraavilla organisaatioille:</p>
      <ul>
          <li>Graves and Sons</li>
          <li>Valdez-Thompson</li>
      </ul>
      
          <p>Erityisviesti: custom message</p>
          
    ''',
  ])
# ---
# name: test_myprofile_creation_email
  list([
    '''
      no-reply@hel.ninja|['jennifer62@example.org']|My profile creation FI|
      <p>Hyvä Kultus ylläpitäjä!</p>
      <p>Uusi palveluntarjoajan tunnus on luotu!</p>
      <address>
          Amanda Newton<br />
          <a href="mailto:hutchinsonrachel@example.org">hutchinsonrachel@example.org</a>
          Käyttäjätunnus: jeffersonkimberly_3MmHFh
      </address>
      <p>Palveluntarjoaja tarvitsee ylläpitäjää (sinua)
      hyväksymään luomansa käyttäjätunnuksen käyttöön:</p>
      <ol>
          <li>Luo puuttuvat organisaatiot LinkedEventsiin</li>
          <li>Luo puuttuvat organisaatiot Kultukseen</li>
          <li>Liitä organisaatiot käyttäjään</li>
          <li>Lisää käyttäjälle staff-lippu, jotta hän saisi oikeudet
          luoda ja muokata tapahtumia.</li>
      </ol>
      <p>Käyttäjä haluaisi edustaa seuraavia organisaatioita:</p>
      <ul>
          <li>Smith, Wood and Baker</li>
          <li>Wolfe, Rogers and Morgan</li>
      </ul>
      <p>Muokataksesi luotua käyttäjätunnusta,
      klikkaa <a href="https://test-domain/admin/organisations/user/123/change/" target="_blank">tästä</a>!</p>
      <p>Nähdäksesi listan käyttäjistä, klikkaa
      <a href="https://test-domain/admin/organisations/user/" target="_blank">tästä</a>.</p>
      
          <p>Erityisviesti: custom message</p>
  
    ''',
    '''
      no-reply@hel.ninja|['lsimmons@example.com']|My profile creation FI|
      <p>Hyvä Kultus ylläpitäjä!</p>
      <p>Uusi palveluntarjoajan tunnus on luotu!</p>
      <address>
          Amanda Newton<br />
          <a href="mailto:hutchinsonrachel@example.org">hutchinsonrachel@example.org</a>
          Käyttäjätunnus: jeffersonkimberly_3MmHFh
      </address>
      <p>Palveluntarjoaja tarvitsee ylläpitäjää (sinua)
      hyväksymään luomansa käyttäjätunnuksen käyttöön:</p>
      <ol>
          <li>Luo puuttuvat organisaatiot LinkedEventsiin</li>
          <li>Luo puuttuvat organisaatiot Kultukseen</li>
          <li>Liitä organisaatiot käyttäjään</li>
          <li>Lisää käyttäjälle staff-lippu, jotta hän saisi oikeudet
          luoda ja muokata tapahtumia.</li>
      </ol>
      <p>Käyttäjä haluaisi edustaa seuraavia organisaatioita:</p>
      <ul>
          <li>Smith, Wood and Baker</li>
          <li>Wolfe, Rogers and Morgan</li>
      </ul>
      <p>Muokataksesi luotua käyttäjätunnusta,
      klikkaa <a href="https://test-domain/admin/organisations/user/123/change/" target="_blank">tästä</a>!</p>
      <p>Nähdäksesi listan käyttäjistä, klikkaa
      <a href="https://test-domain/admin/organisations/user/" target="_blank">tästä</a>.</p>
      
          <p>Erityisviesti: custom message</p>
  
    ''',
    '''
      no-reply@hel.ninja|['patrickkenneth@example.com']|My profile creation EN|
      <p>Dear Kultus Admin!</p>
      <p>A new Kultus provider user profile is created!</p>
      <address>
          Amanda Newton<br />
          <a href="mailto:hutchinsonrachel@example.org">hutchinsonrachel@example.org</a>
          Username: jeffersonkimberly_3MmHFh
      </address>
      <p>The provider who created the user profile needs an admin (you)
      to accept the user profile:</p>
      <ol>
          <li>Create the missing organisations to LinkedEvents</li>
          <li>Create the missing organisations to Kultus</li>
          <li>Link the user to organisations</li>
          <li>Set the staff -flag so the user would receive the permissions
          to create and edit their events.</li>
      </ol>
      <p>The user would like to represent these organisations:</p>
      <ul>
          <li>Smith, Wood and Baker</li>
          <li>Wolfe, Rogers and Morgan</li>
      </ul>
      <p>To edit the newly created user profile,
      click <a href="https://test-domain/admin/organisations/user/123/change/" target="_blank">here</a>!</p>
      <p>To see a full list of users, click
      <a href="https://test-domain/admin/organisations/user/" target="_blank">here</a>.</p>
      
          <p>Custom message: custom message</p>
  
    ''',
  ])
# ---
